### PR TITLE
GH-502 GH-527 Implement dependency injector. Local maven repository.

### PR DESCRIPTION
**Pull request by Rollczi**
Resolve GH-502
### Only do the review in InteliJ IDEA
image

### TL;DR

<details>
<summary>PL</summary>
Automatyczny system wstrzykiwania zależności.

**Minusy**
- Jest więcej magii.
- Mniej kontroli nad kodem w trakcie compiletime.
- Większy potencjał na złe rozkładanie zależności i niestosowanie się do zasady [Dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle)

**Plusy**
- Automatyczna konfiguracja i rejestracja serwisów komend listenerów itd
- Potencjał na dużą hermetyzację package-private w obrębie jednej paczki:
image
- Dobrze rozegrane wyżej ^^^ może pomóc w izolacji zależności i stworzenie punktów granicznych na fasadach aka serwisach więcej info o tym podejściu: https://www.youtube.com/watch?v=ILBX9fa9aJo
</details>

<details>
<summary>EN</summary>
Automatic dependency injection system.

**Cons**
- Code is more "magic"
- You have less control over the code at compile time. 
- You can easily break the [Dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle) principle

**Pros**
- Automatic configuration and registration services, commands, listeners etc.
- Potential for high package-private encapsulation in one package:
image
- If we use package-private encapsulation ^^^ we can separate dependencies and create facades (aka services) as domain boundary points. More about that; https://www.youtube.com/watch?v=ILBX9fa9aJo
</details>

## EternalCore DI
Dependency Injection (DI) is an object-oriented programming design pattern that enhances flexibility, maintainability, and modularity by managing component dependencies. It allows an external source to inject dependencies into a class, reducing tight coupling and improving testability.

Key DI concepts include:
- **Beans** (instances of classes managed by the bean container)
- **Bean container** (stores beans)
- **Dependent class** (those needing dependencies from the bean container)

Used DI type
**Constructor Injection:** Dependencies are injected through the class's constructor (annotated as `@Inject`)

#### 1. Repository
```java
@Repository // <- means a class that will be injected into other classes. (dependency)
public class UserRepositoryImpl implements UserRepository {
    // Implementation
}
```

#### 2. Config
```java
@ConfigurationFile // <- marks a class that will be registered and loaded as a config.
public class UserConfiguration implements ReloadableConfig {
    String example = "example";
    int exampleInt = 1;
    boolean exampleBoolean = true;

    @Override
    public Resource resource(File folder) {
        return Source.of(folder, "user-configuration.yml");
    }
}
```

#### 3. Service / Manager etc.
```java
@Service // <- means a class that will be injected into other classes. (dependency)
public class UserService {
    
    private final UserRepository exampleRepository;
    
    @Inject // <- marks a class as dependent on other classes. (dependent class)
    public UserService(UserRepository exampleRepository) {
        this.exampleRepository = exampleRepository;
    }
    
}
```

#### 4. Bukkit task
```java
@Task(delay = 20L, unit = TimeUnit.SECONDS) // <- marks a class that will be registered as a task.
public class UserTask implements Runnable {
    
    private final UserService exampleService;
    
    @Inject // <- marks a class as dependent on other classes. (dependent class)
    public UserTask(UserService exampleService) {
        this.exampleService = exampleService;
    }
    
    @Override
    public void run() {
        // do something
    }
    
}
```
#### 5. Bukkit listener
```java
@Controller // <- marks a class that will be registered as a bukkit listener.
public class UserListener implements Listener {
    
    private final UserService exampleService;
    
    @Inject // <- marks a class as dependent on other classes. (dependent class)
    public UserListener(UserService exampleService) {
        this.exampleService = exampleService;
    }
    
    @EventHandler
    void onJoin(AsyncPlayerJoinEvent event) {
        // do something
    }
    
}
```
#### 6. Command
```java
@Route(name = "example") // <- marks a class that will be registered as a command.
public class UserCommand {
    
    private final UserService exampleService;
    
    @Inject // <- marks a class as dependent on other classes. (dependent class)
    public UserCommand(UserService exampleService) {
        this.exampleService = exampleService;
    }
    
    @Execute
    void onUserCommand(CommandSender sender, @Arg String arg) {
        // do something
    }
    
}
```

#### 7. Subscriber
```java
@Controller // <- marks a class that will be registered as an event subscriber.
public class UserSubscriber implements Subscriber {
    
    private final UserService exampleService;
    
    @Inject // <- marks a class as dependent on other classes. (dependent class)
    public UserListener(UserService exampleService) {
        this.exampleService = exampleService;
    }
    
    @Subscribe
    void onInitialize(EternalInitializeEvent event) {
        // do something on initialize
    }
    
}
```

#### 8. BeanSetup
BeanSetup is a class that allows you to register beans in the bean container.
It is used to register dependencies that are cannot be registered in the bean container using annotations. e.g. MiniMessage, AdventureProvider, HikariDataSource, etc.
```java

@BeanSetup // <- marks a class that will be registered as a bean holder.
public class UserBeanSetup {

    @Bean // <- marks a method that will be registered as a bean. (dependency)
    public UserProvider exampleProvider() {
        UserProviderImpl exampleProvider = new UserProviderImpl();
        exampleProvider.doSomething();
        
        return new UserProviderImpl();
    }
    
    @Bean // <- marks a method that will be registered as a bean. (dependency)
    public OtherProvider otherProvider(Server server) {
        return new OtherProvider(server);
    }
}
```

### CODE REVIEW

File: eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        this.eternalCoreApi = eternalCoreApi;
+        this.logger = logger;
+        this.checkSoftware();
+    }
+
+    void initialize() {
+        EternalCoreApiProvider.initialize(this.eternalCoreApi);
+
+        long millis = this.stopwatch.elapsed(TimeUnit.MILLISECONDS);
+        this.logger.info("Successfully loaded EternalCore in " + millis + "ms");
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java:26]: I would move Stopwatch to the initialize method.
```suggestion

    EternalCoreEnvironment(EternalCoreApi eternalCoreApi, Logger logger) {
        this.eternalCoreApi = eternalCoreApi;
        this.logger = logger;
        this.checkSoftware();
    }

    void initialize() {
        Stopwatch stopwatch = Stopwatch.createStarted();
        
        EternalCoreApiProvider.initialize(this.eternalCoreApi);

        long millis = stopwatch.elapsed(TimeUnit.MILLISECONDS);
        this.logger.info("Successfully loaded EternalCore in " + millis + "ms");
```

**Review by imDMK — APPROVED** [review]

**Review by vLuckyyy — APPROVED** [review]

**Comment by CitralFlo** [comment]
Please switch to English in first part of the desription.

**Review by Hyd3r1 — APPROVED** [review]

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/gamemode/GameModeConfigurator.java [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
     private final CommandConfiguration commandConfiguration;
 
-    public GameModeConfigurator(CommandConfiguration commandConfiguration) {
+    @Inject
+    GameModeConfigurator(CommandConfiguration commandConfiguration) {
         this.commandConfiguration = commandConfiguration;
     }
 
     @Override
     public State edit(State state) {
```
Comment (vLuckyyy) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/gamemode/GameModeConfigurator.java:19]: Remove public?

Comment (Rollczi) [reply eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/gamemode/GameModeConfigurator.java:19]: hm?

**Review by CitralFlo — APPROVED** [review]

**Review by eripe14 — APPROVED** [review]

**Review by igoyek — APPROVED** [review]

File: eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    void shutdown() {
+        EternalCoreApiProvider.deinitialize();
+    }
+
+    private void checkSoftware() {
+        Environment environment = PaperLib.getEnvironment();
+
+        if (!environment.isSpigot()) {
+            this.logger.warning("Your server running on unsupported software, use spigot/paper minecraft software and other spigot/paper 1.19x forks");
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java:37]:
```suggestion
            this.logger.warning("Your server running on unsupported software, please use Spigot/Paper or their other 1.19+ forks");
```

File: eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            this.logger.warning("We recommend using paper, download paper from https://papermc.io/downloads");
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java:38]:
```suggestion
            this.logger.warning("We recommend using Paper, download it from https://papermc.io/downloads");
```

File: eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            this.logger.warning("WARRING: Supported minecraft version is 1.17-1.19x");
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/EternalCoreEnvironment.java:39]:
```suggestion
            this.logger.warning("WARNING: Supported Minecraft versions are 1.17-1.19x");
```

File: eternalcore-core/src/main/java/com/eternalcode/core/database/DatabaseManagerSetup.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Bean
+    DatabaseManager databaseManager(PluginConfiguration pluginConfiguration, Logger logger, File dataFolder) {
+        DatabaseManager databaseManager = new DatabaseManager(pluginConfiguration, logger, dataFolder);
+
+        try {
+            databaseManager.connect();
+        }
+        catch (SQLException exception) {
+            logger.severe("Can not connect to database! Some functions may not work!");
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/database/DatabaseManagerSetup.java:25]:
```suggestion
            logger.severe("Could not connect to database! Some functions may not work properly!");
```

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 7]
```diff
@@ -13,13 +15,21 @@
...
+class WarpManager {
 
     private final Map<String, Warp> warpMap = new HashMap<>();
     private final WarpRepository warpRepository;
 
+    @Inject
     private WarpManager(WarpRepository warpRepository) {
         this.warpRepository = warpRepository;
+
+        warpRepository.getWarps().thenAcceptAsync(warps -> { //TODO: Use only completable
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:28]: TODO rozumiem że do dokończenia w przyszłym PR?

File: eternalcore-core/src/main/java/com/eternalcode/core/injector/annotations/component/Component.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ * @see Service
+ * @see Repository
+ * @see Task
+ * @see Controller
+ * @see ConfigurationFile
+ * @see BeanSetup
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Component {
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/injector/annotations/component/Component.java:21]: Correct me if I'm wrong, but if I recall correctly, on last meeting we said that we'll consider removing this class.

File: eternalcore-core/src/main/java/com/eternalcode/core/injector/scan/DependencyScannerFactory.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                Task.class,
+                Controller.class,
+                ConfigurationFile.class,
+                BeanSetup.class,
+                Route.class,
+                RootRoute.class,
+                LiteArgument.class,
+                LiteNativeArgument.class,
+                LiteHandler.class,
+                LiteContextual.class
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/injector/scan/DependencyScannerFactory.java:38]: Would be a good practice to separate LiteCommand dependencies from normal ones
```suggestion
                BeanSetup.class,
                Component.class,
                ConfigurationFile.class,
                Controller.class,
                Repository.class,
                Service.class,
                Task.class,
                
                LiteNativeArgument.class,
                LiteHandler.class,
                LiteContextual.class,
                Route.class,
                RootRoute.class
```

File: eternalcore-core/src/main/java/com/eternalcode/core/metrics/BStatsMetricsSetup.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import org.bstats.bukkit.Metrics;
+import org.bukkit.plugin.java.JavaPlugin;
+
+@Controller
+class BStatsMetricsSetup implements Subscriber {
+
+    @Subscribe(EternalInitializeEvent.class)
+    public void onInitialize(JavaPlugin javaPlugin) {
+        Metrics metrics = new Metrics(javaPlugin, 13964);
+        //metrics.addCustomChart(new SingleLineChart("users", () -> 0));
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/metrics/BStatsMetricsSetup.java:16]:
```suggestion
        
```

File: eternalcore-core/src/main/java/com/eternalcode/core/publish/event/EternalEvent.java [hunk 11]
```diff
@@ -0,0 +1,6 @@
+package com.eternalcode.core.publish.event;
+
+import com.eternalcode.core.publish.PublishEvent;
+
+public interface EternalEvent extends PublishEvent {
```
Comment (Jakubk15) [inline eternalcore-core/src/main/java/com/eternalcode/core/publish/event/EternalEvent.java:5]: Change the name to `EternalCoreEvent`

**Review by Jakubk15 — APPROVED** [review]

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java (hunk 7 above)
Comment (Rollczi) [reply eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:28]: yup

File: eternalcore-core/src/main/java/com/eternalcode/core/injector/annotations/component/Component.java (hunk 8 above)
Comment (Rollczi) [reply eternalcore-core/src/main/java/com/eternalcode/core/injector/annotations/component/Component.java:21]: We can keep this annotation for cases as `ProviderImpl`

File: eternalcore-core/src/main/java/com/eternalcode/core/injector/scan/DependencyScannerFactory.java (hunk 9 above)
Comment (Rollczi) [reply eternalcore-core/src/main/java/com/eternalcode/core/injector/scan/DependencyScannerFactory.java:38]: okey, but i will reorder annotations and add missing `LiteArgument.class`

**Review by Osnixer — APPROVED** [review]
